src/Build/BackEnd/BuildManager/BuildManager.cs

```diff
@@ -832,10 +832,12 @@ public void EndBuild()
                 ShutdownConnectedNodes(false /* normal termination */);
                 _noNodesActiveEvent.WaitOne();
 
-                // Wait for all of the actions in the work queue to drain.  Wait() could throw here if there was an unhandled exception
-                // in the work queue, but the top level exception handler there should catch everything and have forwarded it to the
+                // Wait for all of the actions in the work queue to drain.
+                // _workQueue.Completion.Wait() could throw here if there was an unhandled exception in the work queue,
+                // but the top level exception handler there should catch everything and have forwarded it to the
                 // OnThreadException method in this class already.
                 _workQueue.Complete();
+                _workQueue.Completion.Wait();
```

Comment:
Should this statement also be behind the changewave check? (Apologies for not adding this comment earlier.)

Reply:
It is a good point. It was actually bug, I think, `_workQueue.Complete()` do not wait. But it could have, low probability, chance for regresion. Lets put it under ChangeWave.